### ISSUE TITLE
Create an abstraction for stream in realtime

### ISSUE THREAD

**Issue by npawar** [apache/incubator-pinot#2583]
Pinot should be able to support any stream in realtime, not just kafka. Our code currently is not designed to allow that. We have references to Kafka and all realtime code designed assuming the stream is always kafka. 
We will bring in the abstraction for stream. We will do this over multiple iterations. 
Things to be done:
1) Introduce interfaces for Stream, StreamMetadata, StreamProvider, StreamProvider config, factories and others to make realtime code stream agnostic
2) Rename all references to "kafka" appropriately

**Comment by mcvsubbu** [comment]
The things we assume in the code are:

The stream is partitioned (if you want to use a partition-based consumer)
The stream provides APIs to consume all partitions at a time, or a single partition at a time
The stream has a topic name, some timeout settings, etc.
The stream has a (comparable) offset that is a long

Most of the code changes required are moving classes and renaming methods/members to not contain the word 'kafka'. We can do this in a few steps to get to the point where all Kafka related code is in one package, and then look to making that an class(or classes) that can be injected.

The segment completion protocol deals with offsets as being long, so some work may be needed if we wanted a generic Comparable as an offset. We can get there later.

**Comment by npawar** [comment]
Tasks:
- In LLRealtimeDataManager, remove the use of Kafka specific errors
- Introduce abstraction for StreamMetadataProvider, which would answer fetchPartitionCount.

**Comment by EronWright** [comment]
@npawar are you envisioning a plugin system here?  i.e. install a stream provider, configure it, provide a facility for storing state (e.g. offsets), some documentation, etc?

**Comment by mcvsubbu** [comment]
@EronWright what we are trying to do is to unweave Kafka and Pinot.  The idea is that you should be able to inject classes that can parse the configuration that they need and provide functionality to read from an incoming stream. This is ongoing iterative work, but we are getting started.

**Comment by EronWright** [comment]
Thanks @mcvsubbu, by formalizing the effort a bit more, the community could jump in and help out.

**Comment by fx19880617** [comment]
Also mentioned in #3364, there is a backward-incompatible issue of config consumer.factory.class.name. This was not a required config when we create realtime tables.

Could we add a default value for this or infer it from somewhere?

Below are typical configs are are using:
For highlevel consumer:

      "streamConfigs": {
        "stream.kafka.decoder.class.name": "com.uber.uPinot.decoder.HeatpipeKafkaMessageDecoder",
        "streamType": "kafka",
        "stream.kafka.hlc.zk.connect.string": "kafka-zookeeper",
        "realtime.segment.flush.threshold.size": "1000000",
        "stream.kafka.consumer.type": "highLevel",
        "realtime.segment.flush.threshold.time": "18000000",
        "stream.kafka.consumer.prop.auto.offset.reset": "smallest",
        "stream.kafka.topic.name": "some-topic"
      },
For simple consumer

      "streamConfigs": {
        "stream.kafka.decoder.class.name": "com.uber.uPinot.decoder.HeatpipeKafkaMessageDecoder",
        "streamType": "kafka",
        "stream.kafka.hlc.zk.connect.string": "kafka-zookeeper",
        "realtime.segment.flush.threshold.size": "1000000",
        "stream.kafka.consumer.type": "simple",
        "realtime.segment.flush.threshold.time": "18000000",
        "stream.kafka.broker.list": "kafka-broker",
        "stream.kafka.consumer.prop.auto.offset.reset": "largest",
        "stream.kafka.topic.name": "sometopic"
      }

**Comment by fx19880617** [comment]
a hotfix: #3380

**Comment by mcvsubbu** [comment]
@EronWright , @fx19880617  we believe we have  completed work on this PR, and will be able to support streams that look and feel like Kafka :-), but are not Kafka. We will be adding documentation on what we expect the stream should support/provide, and how you should implement a new plug-in. Meanwhile, if you can take a look at StreamConsumerFactoryProvider, that should give you a pretty good idea of what needs to be implemented to support another stream. Please let us know if some abstractions need to move around or do not work for you. Better yet, submit a PR with how you want the abstraction, and we can get to something that works for all of us.

Thanks

**Comment by npawar** [comment]
Here's the wiki for pluggable streams: https://github.com/linkedin/pinot/wiki/Pluggable-Streams-in-Realtime

**Comment by EronWright** [comment]
Thanks @mcvsubbu and @npawar for the ongoing effort.

**Comment by mcvsubbu** [comment]
@EronWright  somehow we completely missed this comment. We apologize for the late response.

The re-factoring for offset not being a 'long' is hard and touches multiple places. The code on the server side (or that on the controller side) alone is probably easier than the idea of keeping it compatible between the two. We need to make sure that the segment completion works right with other types of offset. 

If the offset is a Comparable, then I think it could be a little easier. We just have to try out the re-factor. 

Alternatively, as you say, the high-level interface can be targeted. Two major issues we found with High level streams: 
1. We are limited to consumption rate of one host.  This was a non-starter for some of our high ingestion use cases.
2. Operationally it was a nightmare. One had to be careful about setting the Kafka instance config manually while expanding the cluster or replacing a node if it is down . We had gotten into situation where the Kafka traffic was split between a couple of hosts, leading to wrong results for the application. Addition of a new host to the cluster meant (sometimes) waiting for several days for that host to catch up, meanwhile making sure that new software is not deployed to that host. Some of these can be overcome with additional software, but we decided to go with LLC

Hope this helps.
